### PR TITLE
fix(lsp): inlay hints are rendered in the correct order

### LINKED ISSUE #29647: Wrong ordering of inlay hints with equal position
### Problem

LSP specifies:

> If multiple hints have the same position, they will be shown in the order they appear in the response.

Neovim currently shows inlay hints in different order.

### Steps to reproduce using "nvim -u minimal_init.lua"

No need for `minimal_init.lua`, the issue reproduces with `nvim --clean` by running two commands:

```
:lua vim.lsp.start { name = "r3", cmd = {"python3", "/path/to/repro.py"}, root_dir = "." }
:lua vim.lsp.inlay_hint.enable()
```

where `repro.py` is a minimal language server that sends digits 0 to 4 as four separate inlay hints at position 0: https://gist.github.com/GoldsteinE/702327e04ddc3aedd3e6af7e9a70e68e

Neovim currently shows `04321`.

The issue also reproduces with rust-analyzer, see https://github.com/rust-lang/rust-analyzer/issues/17574 for a minimal config and example.

### Expected behavior

`01234` (in that order) appears at position 0.

### Neovim version (nvim -v)

NVIM v0.11.0-dev-393+g3c5abf01b

### Language server name/version

custom, attached

### Operating system/version

NixOS

### PR DESCRIPTION
**Pull request by amitds1997**
Fixes #29647 

Problem:
When there are multiple inlay hints present at the same position, they should be rendered in the order they are received in the response from LSP as per the LSP spec. Currently, this is not respected.

Reason:
When each extmark call is done, the mark is stored in a MarkTree which keeps track of all marks. When these marks get applied during redraw, the order of access does not preserve the order of insertion ([source](https://github.com/amitds1997/neovim/blob/54b9646210e8bc4889ef7dd45792909630e4ec10/src/nvim/decoration.c?plain=1#L593-L600)). So, adding marks in order of insertion could be other solution too but since `MarkTree` is used for tracking other kind of marks (like signs) I was unsure if such a traversal method makes sense. If that might be a better solution, let me know.

Solution:
Gather all hints for a given position, and then set it in a single extmark call instead of multiple `nvim_buf_set_extmark` calls. This leads to fewer `nvim_buf_set_extmark` calls and correct inlay hints being rendered.

### CODE REVIEW

**Comment by mfussenegger** [comment]
Looks good to me but could you also add a test case?

**Comment by clason** [comment]
(and rebase)

**Comment by amitds1997** [comment]
> Looks good to me but could you also add a test case?

Thanks for the review. Done in the latest push.

File: test/functional/plugin/lsp/inlay_hint_spec.lua [hunk 1]
```diff
@@ -258,3 +258,74 @@ describe('vim.lsp.inlay_hint', function()
     end)
   end)
 end)
+
+describe('Inlay hints are correctly rendered', function()
```
Comment (mfussenegger) [inline test/functional/plugin/lsp/inlay_hint_spec.lua:262]:
```suggestion
describe('Inlay hints handler', function()
```

Correctly is redundant/implied for a test case. See suggestion below for a variant that also communicates the expected outcome

File: test/functional/plugin/lsp/inlay_hint_spec.lua [hunk 2]
```diff
@@ -258,3 +258,74 @@ describe('vim.lsp.inlay_hint', function()
...
+    vim.api.nvim_win_set_buf(0, bufnr)
+
+    client_id = vim.lsp.start({ name = 'dummy', cmd = server.cmd })
+  ]],
+      response
+    )
+    insert(text)
+  end)
+
+  it('when multiple inlay hints are at same character position', function()
```
Comment (mfussenegger) [inline test/functional/plugin/lsp/inlay_hint_spec.lua:321]:
```suggestion
  it('renders hints with same position in received order, function()
```

**Comment by mfussenegger** [comment]
Thanks for the test case addition. I left a minor comment regarding naming, otherwise lgtm

File: test/functional/plugin/lsp/inlay_hint_spec.lua (hunk 1 above)
Comment (amitds1997) [reply test/functional/plugin/lsp/inlay_hint_spec.lua]: Makes sense. 
Done in the latest commit: [7248b73](https://github.com/neovim/neovim/pull/29707/commits/7248b7363ffdd00e0ac3288a92d4c0d2f7396a5c). Also, rebased with current master.

File: test/functional/plugin/lsp/inlay_hint_spec.lua (hunk 2 above)
Comment (amitds1997) [reply test/functional/plugin/lsp/inlay_hint_spec.lua]: Done in [7248b73](https://github.com/neovim/neovim/pull/29707/commits/7248b7363ffdd00e0ac3288a92d4c0d2f7396a5c).

**Comment by clason** [comment]
Backport?

**Comment by MariaSolOs** [comment]
> Backport?

I think we should.

**Pull request merged**